tla/consensus/abs.tla

```diff
@@ -37,22 +34,65 @@ Copy(i) ==
         /\ \E l \in 1..(Len(cLogs[j]) - Len(cLogs[i])) : 
                 cLogs' = [cLogs EXCEPT ![i] = @ \o SubSeq(cLogs[j], Len(@) + 1, Len(@) + l)]
 
-\* A node i with the longest log can extend its log upto length k.
-Extend(i, k) ==
+\* A node i with the longest log can non-deterministically extend
+\* its log by any finite number of log entries.  An implementation
+\* may choose a particular number of log entries by which to extend
+\* the log to prevent the leader from racing ahead of the followers.
+Extend(i) ==
     /\ \A j \in Servers : Len(cLogs[j]) \leq Len(cLogs[i])
-    /\ \E l \in 0..(k - Len(cLogs[i])) : 
-        \E s \in [1..l -> Terms] :
+    /\ \E s \in Seq(Terms) :
             cLogs' = [cLogs EXCEPT ![i] = @ \o s]
 
-ExtendToMax(i) == Extend(i, MaxLogLength)
+ExtendAxiom(i) ==
+    \* i has the longest log.
+    /\ \A j \in Servers : Len(cLogs[j]) \leq Len(cLogs[i])
+    \* cLogs remains a function mapping from servers to logs.
+    /\ cLogs' \in [Servers -> Seq(Terms)]
+    \* i *extends* its log
+    /\ IsPrefix(cLogs[i], cLogs'[i])
+    \* The other logs remain the same.
+    /\ \A j \in Servers \ {i} : cLogs'[j] = cLogs[j]
+
+\* Extend and ExtendAxiom are logically equivalent definitions.  However,
```

Comment:
What is the value of keeping two sets of equivalent definitions, one of which is more efficient? Why don't we just keep the efficient implementations?

Reply:
I believe the original definitions (`Extend` and `CopyMaxAndExtend`) are still used by `MCabs.tla`. If my understanding is correct the original definitions are in the usual format to "generate" (big air quotes) next states but the `Axiom` versions are for checking behaviors for refinement

Reply:
Ok, I see now. It's unfortunate that there isn't a way to compose the "generation" and the axiom, rather have two and need to check their equivalence, but I guess it can't be helped.

Reply:
The "axiomatic" definitions are cheaper to evaluate if TLC has to check if an action (a pair of states) is true.  This is the case when TLC checks refinement (of a spec without a fairness constraint).  The original definitions are cheaper to evaluate when TLC generates the successor states.  This is the case during ordinary model checking of e.g. `abs` itself.